rp2040-hal/src/spi.rs

```diff
@@ -78,53 +78,131 @@ impl<S: State, D: SpiDevice, const DS: u8> Spi<S, D, DS> {
-        // post-divide. Prescale is an even number from 2 to 254 inclusive.
-        for prescale_option in (2u32..=254).step_by(2) {
-            // We need to use an saturating_mul here because with a high baudrate certain invalid prescale
-            // values might not fit in u32. However we can be sure those values exeed the max sys_clk frequency
-            // So clamping a u32::MAX is fine here...
-            if freq_in < ((prescale_option + 2) * 256).saturating_mul(baudrate) {
-                prescale = prescale_option as u8;
-                break;
-            }
-        }
+
+        let settings = calc_spi_clock_divider_settings_for_baudrate(freq_in, baudrate);
 
         // We might not find a prescale value that lowers the clock freq enough, so we leave it at max
-        debug_assert_ne!(prescale, u8::MAX);
-
-        // Find largest post-divide which makes output <= baudrate. Post-divide is
-        // an integer in the range 0 to 255 inclusive.
-        for postdiv_option in (1..=255u8).rev() {
-            if freq_in / (prescale as u32 * postdiv_option as u32) > baudrate {
-                postdiv = postdiv_option;
-                break;
-            }
-        }
+        debug_assert_ne!(settings.prescale, u8::MAX);
+
+        self.set_baudrate_from_settings(&settings);
+
+        // Return the frequency we were able to achieve
+        use fugit::RateExtU32;
+        (freq_in / (settings.prescale as u32 * (1 + settings.postdiv as u32))).Hz()
+    }
 
+    /// Set the baudrate using a previously calculated [SpiClockDividerSettings]
+    pub fn set_baudrate_from_settings(&mut self, settings: &SpiClockDividerSettings) {
         self.device
             .sspcpsr
-            .write(|w| unsafe { w.cpsdvsr().bits(prescale) });
+            .write(|w| unsafe { w.cpsdvsr().bits(settings.prescale) });
         self.device
             .sspcr0
-            .modify(|_, w| unsafe { w.scr().bits(postdiv) });
+            .modify(|_, w| unsafe { w.scr().bits(settings.postdiv) });
+    }
 
-        // Return the frequency we were able to achieve
-        use fugit::RateExtU32;
-        (freq_in / (prescale as u32 * (1 + postdiv as u32))).Hz()
+    /// Set the mode
+    ///
+    /// Momentarily disables / enables the device so be careful of truncating ongoing transfers.
+    pub fn set_mode(&mut self, mode: Mode) {
+        // disable the device
+        self.device.sspcr1.modify(|_, w| w.sse().clear_bit());
+
+        // Set the polarity and phase
+        self.device.sspcr0.modify(|_, w| {
+            w.spo()
+                .bit(mode.polarity == Polarity::IdleHigh)
+                .sph()
+                .bit(mode.phase == Phase::CaptureOnSecondTransition)
+        });
+
+        // enable the device
+        self.device.sspcr1.modify(|_, w| w.sse().set_bit());
+    }
+
+    /// Wait until all operations have completed and the bus is idle.
+    pub fn flush(&self) -> Result<(), nb::Error<Infallible>> {
```

Comment:
Another remark: The future of `nb` is a little bit unclear. It might get replaced by an `async` API.
While I'd say that we should continue supporting the `nb` traits of embedded-hal for the foreseeable future, I wouldn't add methods using it to the the rp2040-hal impls.

So this method should have the same signature as https://docs.rs/embedded-hal/1.0.0-alpha.9/embedded_hal/spi/trait.SpiBusFlush.html#tymethod.flush, `fn flush(&mut self) -> Result<(), Self::Error>`.

Reply:
`Self::Error` is [currently being implemented](https://github.com/rp-rs/rp-hal/blob/2fb3e2a7e96cc9901d38c24b4a60b9960d90e866/rp2040-hal/src/spi.rs#L230) as `Infallable`, so the proposed signature change would mean `flush` would _have_ to be blocking, right?

```rust
    fn flush(&mut self) -> Result<(), Self::Error> {
        while self.device.sspsr.read().bsy().bit() {}
        Ok(())
    }
```

If that's the case, why not make the inherent impl return `()` and allow an eh alpha `SpiBusFlush` implementation to use that.  See[ _Fallability_ discussion](https://github.com/rust-embedded/embedded-hal/blob/74f9b9bd600f195456f6fe588fbc8f59b02b3894/MIGRATING-0.2-1.0.md#fallibility) in proposed migration guide.

```rust
    fn flush(&mut self) {
        while self.device.sspsr.read().bsy().bit() {}
    }
```

Or for that matter, if it is helpful to have an inherent non-blocking implementation (I honestly don't know if that's helpful) why not let the inherent impl be non-blocking (keep the current signature with `Result<(), nb::Error<Infallible>>`) to be consistent with the other inherent impls and the `SpiBusFlush` trait impl loop over it to make it blocking.

Reply:
Just realized this PR is based on a relatively old commit -- this needs a rebase. 

So if rebased, `Spi<Enabled, D, DS>` now has an `is_busy` method which can be used to implement flush. I think it makes sense to only implement `flush` for enabled SPI anyway.

```rust
impl<D: SpiDevice, const DS: u8> Spi<Enabled, D, DS> {
    /* ... */
    fn flush(&self) {
        while self.is_busy {}
    }
}
```

which would allow later:

```rust
    #[cfg(feature = "eh1_0_alpha")]
    impl<D: SpiDevice> eh1::SpiBusFlush for Spi<Enabled, D, $nr> {
        fn flush(&mut self) -> Result<(), Self::Error> {
            self.flush();
            Ok(())
        }
    }
```

Reply:
> So if rebased, `Spi<Enabled, D, DS>` now has an `is_busy` method which can be used to implement flush.

Exactly, and `is_busy` is functionally equivalent to the non-blocking version of `flush`, so it's sufficient to have the blocking `flush` method.